src/main/java/com/contentgrid/gateway/security/jwt/issuer/PropertiesBasedJwtClaimsSigner.java

```diff
@@ -0,0 +1,129 @@
+import com.nimbusds.jose.produce.JWSSignerFactory;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+import java.util.stream.Stream;
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.support.ResourcePatternResolver;
+
+@RequiredArgsConstructor
+public class PropertiesBasedJwtClaimsSigner implements JwtClaimsSigner {
+    private final JwtClaimsSignerProperties properties;
+    private final ResourcePatternResolver resourcePatternResolver;
+    private final JWSSignerFactory jwsSignerFactory;
+    private final Random random;
+
+    public PropertiesBasedJwtClaimsSigner(JwtClaimsSignerProperties properties, ResourcePatternResolver resourcePatternResolver) {
+        this(properties, resourcePatternResolver, new DefaultJWSSignerFactory(), new Random());
+    }
+
+    public interface JwtClaimsSignerProperties {
+        String getActiveKeys();
+        String getAllKeys();
+        Set<JWSAlgorithm> getAlgorithms();
+    }
+
+    @SneakyThrows
+    private static JWK createFromSigningKey(Resource resource) {
+        var jwk = JWK.parseFromPEMEncodedObjects(resource.getContentAsString(StandardCharsets.UTF_8));
+        if(jwk instanceof RSAKey rsaKey) {
+            return new RSAKey.Builder(rsaKey)
+                    .keyIDFromThumbprint()
+                    .keyUse(KeyUse.SIGNATURE)
+                    .build();
+        } else if(jwk instanceof ECKey ecKey) {
+            return new ECKey.Builder(ecKey)
+                    .keyIDFromThumbprint()
+                    .keyUse(KeyUse.SIGNATURE)
+                    .build();
+        } else if(jwk instanceof OctetKeyPair octetKeyPair) {
+            return new OctetKeyPair.Builder(octetKeyPair)
+                    .keyIDFromThumbprint()
+                    .keyUse(KeyUse.SIGNATURE)
+                    .build();
+        } else {
+            throw new IllegalArgumentException("Unsupported JWK key type %s; use RSA, EC or OKP".formatted(jwk.getKeyType()));
+        }
+    }
+
+    @SneakyThrows
+    private Stream<JWK> createSigningKeysFromPath(String path) {
+        if(path == null) {
+            return Stream.empty();
+        }
+        return Arrays.stream(resourcePatternResolver.getResources(path))
+                .map(PropertiesBasedJwtClaimsSigner::createFromSigningKey);
+    }
+
+    private List<JWK> getActiveSigningKeys() {
+        return createSigningKeysFromPath(properties.getActiveKeys())
+                .toList();
+    }
+
+    private List<JWK> getAllSigningKeys() {
+        return Stream.concat(
+                getActiveSigningKeys().stream(),
+                createSigningKeysFromPath(properties.getAllKeys())
```

Comment:
so properties.all-keys should not include the active-keys (or you would have duplicates here) ?
in that case, should we come up with a better name ?

Reply:
It is fine to have duplicates in here